docs: Add details to New Documentation Site Engine project (#167)
keptn/community: mentorship/gsoc/2022/projects/new-docs-site-engine/README.md

## Acknowledgments

I am deeply grateful to my mentors [Indermohan Singh](https://github.com/imsingh), [Meg McRoberts](https://github.com/StackScribe), and [Suraj Banakar](https://github.com/vadasambar), who guided me patiently through the project and were always ready to help, review work in progress, gave feedback regarding the overall shape of the project and the feasibility of certain solutions and maintained a very friendly, supportive, and stimulating atmosphere on the team. They devoted a lot of time to this project and led the video meetings in a well-organized, yet non-stressful manner. Their support and positivity helped me through this project and made me enjoy every bit of it. I would also like to thank [Moritz Wiesinger](https://github.com/mowies), [Oleg Nenashev](https://github.com/oleg-nenashev), and other keptn community members for their support and feedback. Last but not least, I would like to thank Google and Google Summer of Code organizers for this initiative.

## References

- [Contributor GSoC 2022 repository](https://github.com/iamrajiv/GSoC-2022)

> - [Contributor Proposal link](https://summerofcode.withgoogle.com/media/user/f68762051013/proposal/gAAAAABjGvyZjqW5uiTvSpQ-s46PD6X5k6KtZzJ-9x-PSrVRbNh-4k_mTfVrlZgMKJ7iLBYynUjHJ2pahKD4Su86Xch-Caej8cdp7MOQh5JHHXpiNsoTZDk=.pdf)

Reviewer: this link doesn't open for me

- [Acknowledgments](#acknowledgments)
- [References](#references)

## Introduction

This project is about a New Documentation Site Engine for Keptn which can handle a large number of documentation projects of Keptn. Currently, the Keptn documentation site is based on a [Hugo](https://gohugo.io/) theme called [hugo-serif-theme](https://github.com/zerostaticthemes/hugo-serif-theme). For tutorials, [Codelabs](https://github.com/googlecodelabs/tools) is also used. While extremely handy for smaller projects and initial versions of Keptn, this toolchain has serious deficiencies for larger-scale projects like Keptn.

The new Documentation Site engine will be based on Docusaurus. Docusaurus is an optimized site generator in React. Docusaurus helps you to move fast and write content. Build documentation websites, blogs, marketing pages, and more. The main focus of Keptn New Documentation Site Engine will be managing large-scale documentation and complex documentation projects. So two major things which are needed in the documentation engine are the ability to build the documentation from the source located in multiple repositories and support for document versioning.

> Keptn is an event-driven orchestration engine for cloud-native apps. Started in 2018, it uses a declarative approach to automate delivery and operations so they can be scaled up to a large number of services. Keptn evaluates _Service Level Indicators (SLOs)_ and provides a dashboard, alerts, and auto-remediation for them. It also provides integrations with the Cloud Native ecosystem including _Prometheus_, _ArgoCD_, _CloudEvents_, _Backstage_, and many other cloud-native projects.

Reviewer 2: ```suggestion
Keptn is an event-driven orchestration engine for cloud-native apps. Started in 2018, it uses a declarative approach to automate delivery and operations so they can be scaled up to a large number of services. Keptn evaluates _Service Level Indicators (SLIs)_ against the _Service Level Objectives_ (SLOs) and provides a dashboard, alerts, and auto-remediation for them. It also provides integrations with the Cloud Native ecosystem including _Prometheus_, _ArgoCD_, _CloudEvents_, _Backstage_, and many other cloud-native projects.
```

    > I used the `npx create-docusaurus@latest new-keptn-docs-engine classic` command to create the new Docusaurus site so in this case, I used the `classic` template. Docusaurus recommend the `classic` template so that we can get started quickly, and it contains features found in Docusaurus 1. The `classic` template contains `@docusaurus/preset-classic` which includes standard documentation, a blog, custom pages, and a CSS framework (with dark mode support). We can get up and running extremely quickly with the classic template and customize things later on when you have gained more familiarity with Docusaurus. We can also use the template's TypeScript variant by passing the `--typescript` flag. See [TypeScript support](https://docusaurus.io/docs/typescript-support) for more information.

  - [x] Deploying the new Docusaurus site engine to Netlify.

    > [https://github.com/keptn-sandbox/new-keptn-docs-engine](https://github.com/keptn-sandbox/new-keptn-docs-engine) has been deployed to Netlify at [https://keptn-experimental-docs-site.netlify.app/](https://keptn-experimental-docs-site.netlify.app/).

>   - [x] Adding `CONTRIBUTING.md` file with content related to how to add contents in Docusaurus and how to use it concerning Keptn. Also, this work will go on since we will be adding more content in the future according to documentation needs.

Reviewer 2: The "Built with" list in CONTRIBUTING.md should be updated  to include the other things you have implemented - Lighthouse CI, Vale linter (currently using Google style guide), etc.  No need to fill in details about how to use them and such -- just a chance to highlight all the stuff you got working in this project.

    > PR related to adding `CONTRIBUTING.md` file is [https://github.com/keptn-sandbox/new-keptn-docs-engine/pull/4](https://github.com/keptn-sandbox/new-keptn-docs-engine/pull/4).

Issue link: [https://github.com/keptn-sandbox/new-keptn-docs-engine/issues/2](https://github.com/keptn-sandbox/new-keptn-docs-engine/issues/2)

---

> - [x] Adding UI and UX changes, moving the content and search functionality to the new documentation site engine.

Reviewer 2: Why is this marked as completed?  It is also listed in the work to be completed later.  We do not even have the structure defined correctly -- the sections that are under 0.17.x, 0.18.x, etc in the old docs should now be at the same level as "Concepts" and "Installation".

  - [x] Implementing UI and UX changes in the documentation engine like creating components, templates, standalone pages, etc.

    > Created the new components, templates, standalone pages, etc for the documentation engine. I had refactored the default components and templates of the Docusaurus site which were added when the Docusaurus site was initialized.

    > Above task is done and I have pushed the changes directly to the repository. Commit can be found here [https://github.com/keptn-sandbox/new-keptn-docs-engine/commit/b749bbdfebfeaa41c789cd4c91ac29032ac1bbfa](https://github.com/keptn-sandbox/new-keptn-docs-engine/commit/b749bbdfebfeaa41c789cd4c91ac29032ac1bbfa).

    > After getting feedback from mentors on the new documentation site engine, I have refactored the documentation engine. PR can be found here [https://github.com/keptn-sandbox/new-keptn-docs-engine/pull/9](https://github.com/keptn-sandbox/new-keptn-docs-engine/pull/9).

  - [x] Finalizing the things that need to be copied from the current documentation engine to the new documentation site engine with proper structure.

>   - [x] Copying content from the current documentation engine to the new documentation site engine which has been finalized.

Reviewer 2: Why is this marked as completed?

    > As of now not all contents are copied from the current documentation engine to the new documentation site engine instead, some documents are copied to the new documentation site engine.

  - [x] Adding documentation search functionality.

    > As of now I have added local search support using a [docusaurus-lunr-search](https://github.com/praveenn77/docusaurus-lunr-search) local search plugin but in the future, we can update this to [Algolia DocSearch](https://docsearch.algolia.com/).

    > Adding Algolia DocSearch support to the documentation engine is not part of this proposal but it is a suggestion to add it in the future.

    > PR related to adding documentation search functionality can be found here [https://github.com/keptn-sandbox/new-keptn-docs-engine/pull/11](https://github.com/keptn-sandbox/new-keptn-docs-engine/pull/11).

Issue link: [https://github.com/keptn-sandbox/new-keptn-docs-engine/issues/5](https://github.com/keptn-sandbox/new-keptn-docs-engine/issues/5)

---

- [x] Adding Versioning of docs support feature.

> In current docs, we don't have versioning of docs support feature. As of now, we were versioning the docs manually. To solve this problem we have added versioning of the docs support feature using Docusaurus. We can use the versioning CLI to create a new documentation version based on the latest content in the `docs` directory. That specific set of documentation will then be preserved and accessible even as the documentation in the `docs` directory continues to evolve.

Reviewer 3: > We can use the versioning CLI to create a new documentation version

What does versioning CLI mean here? Can you clarify this in the report?

> PR related to adding versioning support can be found here [https://github.com/keptn-sandbox/new-keptn-docs-engine/pull/10](https://github.com/keptn-sandbox/new-keptn-docs-engine/pull/10).

---

- [ ] Adding Multiple repository docs support feature.

This feature is partially done and we want more clarity on this feature. Earlier we thought adding Docusaurus Multi-instance support will solve this problem but it is not the case. The thing is using Docusaurus Multi-instance we can have multiple `docs` folders in the same Docusaurus site and for every `docs` folder, we can put our documentation accordingly. But we wanted something automatic stuff so that we can pull docs from other repository and organize it in these `docs` folder instances.

I got clarity about this feature a little late so I was not able to implement this feature completely.

Some reasons why we want the Multiple repository docs support feature:

> - Keptn CLI docs are generated using the `keptn generate docs` command. As of now, we run it in GitHub action [here](https://github.com/keptn/keptn.github.io/blob/9c649b434c9081028e0cc3535c9fa3a4217941fb/.github/workflows/auto-update.yml#L72) to populate our docs repo with CLI docs.

Reviewer 2: ```suggestion
- The source for the Keptn CLI reference pages is moved from the software repository to the docs repository as a batch job. As of now, we run it in GitHub action [here](https://github.com/keptn/keptn.github.io/blob/9c649b434c9081028e0cc3535c9fa3a4217941fb/.github/workflows/auto-update.yml#L72) to populate our docs repo with CLI docs.
```

> - If we look at the architecture doc [here](https://keptn.sh/docs/concepts/architecture/), the Keptn control plan consists of multiple services which have their READMEs in the main Keptn repository ([example](https://github.com/keptn/keptn/tree/master/jmeter-service)). Just look for any directory with `service` in [https://github.com/keptn/keptn](https://github.com/keptn/keptn), we can see individual READMEs for all such services. It's a good idea to move these READMEs to the docs site because the people who develop and maintain these services are responsible for updating the READMEs. Having them in a single place just makes it easier to update the READMEs and read them later as well.

Reviewer 2: ```suggestion
- If we look at the architecture doc [here](https://keptn.sh/docs/concepts/architecture/), the Keptn control plan consists of multiple services which have their READMEs in the main Keptn repository ([example](https://github.com/keptn/keptn/tree/master/jmeter-service)). Just look for any directory with `service` in [https://github.com/keptn/keptn](https://github.com/keptn/keptn), we can see individual READMEs for all such services. It's a good idea to move these READMEs to the docs site because the people who develop and maintain these services are responsible for updating the READMEs. Having them in a single place just makes it easier to update the READMEs and read them later as well.
We have other files in the software repositories that should be imported to the documentation, such as the `values.yaml` files that configure Helm charts.  We need to define all such entities then devise schemes that allow us to author content in the software repos but build an integrated doc set with that content.
```

Something like this:

```go
const gendocFrontmatterTemplate = `---
sidebar_position: %d
title: "%s"
description: "%s
keywords: "%s"
slug: "%s"
tags: "%s"
---
`
```

We wanted frontmatter like this because it enhances the SEO of docs. But to do this we had to update the go code in [https://github.com/keptn/keptn/blob/master/cli/cmd/generate_docs.go](https://github.com/keptn/keptn/blob/master/cli/cmd/generate_docs.go) file but that will break the current documentation generation process so had to drop this approach.

- Using [docusaurus-plugin-remote-content](https://github.com/rdilweb/docusaurus-plugin-remote-content)

Its Docusaurus plugin downloads content from remote sources. With this plugin, we can write the Markdown for our content somewhere else, and use them on your Docusaurus site, without copying and pasting. This is something I am working on currently and trying to combine with GitHub action too.

> Since the `Adding Multiple repository docs support feature` task was taking a lot of time so I moved to other work and I thought I will pick this up at last because this work needs a lot of time and research and I was not sure how to implement this feature.

> **NOTE:** Multiple repository docs support feature is not yet implemented completely and it is still in progress. As of now `Docs Multi-instance` support is added which is a`@docusaurus/plugin-content-docs` plugin. Still, we want some custom GitHub Action or automation to pull docs from multiple repositories and organize them in a new documentation site engine.

Reviewer 2: Since this is a critical feature that is not implemented, we need Github issues for the specific tasks that are required to get this implemented.  Each one should be targeted and specific and should include any details you have about the work that is required.  The issues should reflect the possibility that different people might work on different issues.  I recommend tagging them with something like "multi-repo support" so we can easily pull them all up for review.

---

- [ ] Adding UI and UX changes, moving the content and search functionality to the new documentation site engine.

>   - [x] Implementing UI and UX changes in the documentation engine like creating components, templates, standalone pages, etc.

Reviewer 2: So you are saying that these are actually done?  I'm not sure I understand how they could be verified without the UI/UX design created and implemented and with virtually none of the core documents transferred to the new repo but perhaps it is possible.
